components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java

```diff
@@ -358,6 +361,78 @@ public AuthenticationResponse authenticate(MessageContext synCtx) {
         }
     }
 
+    private void validateAPIKeyRestrictions(JWTClaimsSet payload, MessageContext synCtx) throws APISecurityException {
+        org.apache.axis2.context.MessageContext axis2MessageContext = ((Axis2MessageContext) synCtx).
+                getAxis2MessageContext();
+
+        String permittedIPList = null;
+        if (payload.getClaim(APIConstants.JwtTokenConstants.PERMITTED_IP) != null) {
+            permittedIPList = (String) payload.getClaim(APIConstants.JwtTokenConstants.PERMITTED_IP);
+        }
+
+        if (StringUtils.isNotEmpty(permittedIPList)) {
+            // Validate client IP against permitted IPs
+            String clientIP = GatewayUtils.getIp(axis2MessageContext);
+
+            if (StringUtils.isNotEmpty(clientIP)) {
+                for (String restrictedIP : permittedIPList.split(",")) {
+                    if (APIUtil.isIpInNetwork(clientIP, restrictedIP.trim())) {
+                        // Client IP is allowed
+                        return;
+                    }
+                }
+                if (log.isDebugEnabled()) {
+                    String apiContext = (String) synCtx.getProperty(RESTConstants.REST_API_CONTEXT);
+                    String apiVersion = (String) synCtx.getProperty(RESTConstants.SYNAPSE_REST_API_VERSION);
+
+                    if (StringUtils.isNotEmpty(clientIP)) {
+                        log.debug("Invocations to API: " + apiContext + ":" + apiVersion +
+                                " is not permitted for client with IP: " + clientIP);
+                    }
+                }
+                throw new APISecurityException(APISecurityConstants.API_AUTH_FORBIDDEN,
+                        "Access forbidden for the invocations");
+            }
+
+        }
+
+        String permittedRefererList = null;
+        if (payload.getClaim(APIConstants.JwtTokenConstants.PERMITTED_REFERER) != null) {
+            permittedRefererList = (String) payload.getClaim(APIConstants.JwtTokenConstants.PERMITTED_REFERER);
+        }
+
+        if (StringUtils.isNotEmpty(permittedRefererList)) {
+            // Validate http referer against the permitted referrers
+            TreeMap<String, String> transportHeaderMap = (TreeMap<String, String>)
+                    axis2MessageContext.getProperty
+                            (org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+            if (transportHeaderMap != null) {
+                String referer = transportHeaderMap.get("Referer");
+                if (StringUtils.isNotEmpty(referer)) {
+                    for (String restrictedReferer : permittedRefererList.split(",")) {
+                        String restrictedRefererRegExp = restrictedReferer.trim()
+                                .replace("*", "[^ ]*");
```

Comment:
can't we move this to the previous line? does it exceed 120 line length?

Reply:
Fixed in https://github.com/wso2/carbon-apimgt/pull/9250